### PR TITLE
Upload coverage to codecov.io is flaky, so do not fail build on uploa…

### PR DESCRIPTION
**Pull request by mavasani**
…d failure

We are seeing lot of failures in debug queue due to codecov upload failures.

### CODE REVIEW

**Review by jmarolf — APPROVED** [review]

File: azure-pipelines.yml
Comment (mavasani) [inline azure-pipelines.yml:44]:
```suggestion
      condition: and(always(), eq(variables['_codeCoverage'], True)) # Upload coverage to codecov.io is flaky, so do not fail build on upload failure
```

**Comment by sharwell** [comment]
Are we using the latest version of the uploader? Curious what the error(s) are.

File: azure-pipelines.yml
Comment (Evangelink) [inline azure-pipelines.yml:44]: Is it possible to introduce some retry pattern?

Comment (mavasani) [reply azure-pipelines.yml:44]: I don't see a reason for it - if someone really wants to look at codecov data and upload failed, they can manually re-run the debug queue on the PR. I don't believe queue should fail if upload of such additional information fails. We already take this approach if publishing test results fail, so this is identical.

**Comment by mavasani** [comment]
> Are we using the latest version of the uploader? Curious what the error(s) are.

I am not sure, but regardless, I don't believe this should be a required gate. It should be optional, just like publishing test results.

**Review by sharwell — CHANGES_REQUESTED** [review]
Can we try updating the codecov uploader tool first?

File: azure-pipelines.yml
Comment (Evangelink) [reply azure-pipelines.yml:44]: I agree with the fact it should not fail the build, I just think that the coverage metric is useful so having a retry seemed like a good option.

**Comment by sharwell** [comment]
> We are seeing lot of failures in debug queue due to codecov upload failures.

I see a few failures (6% of total for the month), but they tend to be bucketed together occurring a few weeks apart. It seems more like a case where we'd work with the partner to improve the upload reliability.

**Comment by jmarolf** [comment]
While I agree with the sentiment @sharwell, we should update the uploader and we should engage to see if we can make things more resilient. I do not think we should block this PR. Unless we are prepared to say that code coverage data is important enough to fail the build, logicallyy it should be optional.

> In the past, this led to extremely long delays between when a configuration error led to continuous upload failures, and when they were found/fixed.

Still don't see this as being useful. We should investigate warnings in the builds, but I think blocking all codeflow for this is overkill.

**Comment by mavasani** [comment]
> In the past, this led to extremely long delays between when a configuration error led to continuous upload failures, and when they were found/fixed. Required uploads tend to provide better long-term build stability.

Lets try to get whatever enforcement/requirement model we want for codecov to flow from arcade. If all arcade based repos mandate this, I am fine to do so for this repo. Otherwise, we are just losing productivity here.

**Comment by sharwell** [comment]
@mavasani can we make it a separate build job (e.g. what dotnet/roslyn#47249 is doing? That would allow two things:

1. We could make it a GitHub check that is not required
2. We could retry _just the upload_ without forcing a new build

**Comment by mavasani** [comment]
@sharwell Those all seem good long term improvements. This PR is aimed at unblocking unnecessary build failures. Feel free to file a follow-up issue to track these improvements.

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn-analyzers/issues/4261 for long term improvements for codecov. I am merging this PR to unblock..

**Pull request merged**
